[22.01] Write to exit code file before running metadata script

mvdbeek opened the pull request:
We need the exit code in the metadata script to figure out the correct
job messages. This would fix
```
FAILED test/functional/test_toolbox_pytest.py::test_tool[unicode_stream_test_2]
FAILED test/functional/test_toolbox_pytest.py::test_tool[strict_shell_test
FAILED test/functional/test_toolbox_pytest.py::test_tool[detect_errors_test_1]
FAILED test/functional/test_toolbox_pytest.py::test_tool[detect_errors_test_2]
FAILED test/functional/test_toolbox_pytest.py::test_tool[detect_errors_test_3]
FAILED test/functional/test_toolbox_pytest.py::test_tool[detect_errors_test_5]
FAILED test/functional/test_toolbox_pytest.py::test_tool[detect_errors_test_9]
FAILED test/functional/test_toolbox_pytest.py::test_tool[detect_errors_test_10]
FAILED test/functional/test_toolbox_pytest.py::test_tool[collection_creates_list_fail_test_1]
```
that are broken when using `metadata_strategy: extended` (https://github.com/galaxyproject/galaxy/issues/13551).
This broke in https://github.com/galaxyproject/galaxy/pull/12459 because we started importing the job back in in extended metadata mode. 

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

mvdbeek committed:
Write to exit code file before running metadata script

We need the exit code in the metadata script to figure out the correct
job messages. This would fix
```
FAILED test/functional/test_toolbox_pytest.py::test_tool[detect_errors_test_1]
FAILED test/functional/test_toolbox_pytest.py::test_tool[detect_errors_test_2]
FAILED test/functional/test_toolbox_pytest.py::test_tool[detect_errors_test_3]
FAILED test/functional/test_toolbox_pytest.py::test_tool[detect_errors_test_5]
FAILED test/functional/test_toolbox_pytest.py::test_tool[detect_errors_test_9]
FAILED test/functional/test_toolbox_pytest.py::test_tool[detect_errors_test_10]
```
that are broken when using `metadata_strategy: extended` (https://github.com/galaxyproject/galaxy/issues/13551).

nsoranzo commented on lib/galaxy/jobs/command_factory.py:
```diff
@@ -290,10 +290,12 @@ def capture_stdout_stderr(self, stdout_file, stderr_file):
             self.return_code_captured = True
             self.append_command(CAPTURE_RETURN_CODE)
+            if job_id:
+                self.append_command(f"echo $return_code > galaxy_{job_id}.ec")
```
Should we remove this line then? https://github.com/galaxyproject/galaxy/blob/release_22.01/lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh#L43

mvdbeek replied:
Probably. I was a little worried that job runners may do something unexpected with the `exit_code_file`, but it doesn't look like that's happening.

mvdbeek committed:
Drop redundant exit_code_path writing from DEFAULT_JOB_FILE_TEMPLATE.sh

mvdbeek committed:
Fix path to exit_code file

nsoranzo commented on lib/galaxy/jobs/command_factory.py:
```diff
@@ -114,7 +114,8 @@ def build_command(
+    commands_builder.capture_return_code(join(working_directory, f"galaxy_{job_wrapper.job_id}.ec"))
 
     if include_metadata and job_wrapper.requires_setting_metadata:
         working_directory = remote_job_directory or job_wrapper.working_directory
```
```suggestion
```

nsoranzo commented on test/unit/app/jobs/test_command_factory.py:
```diff
@@ -37,59 +38,59 @@ def tearDown(self):
         dep_commands = [". /opt/galaxy/tools/bowtie/default/env.sh"]
         self.job_wrapper.dependency_shell_commands = dep_commands
-        self.__assert_command_is(_surround_command(f"{dep_commands[0]}; {MOCK_COMMAND_LINE}; return_code=$?"))
+        self.__assert_command_is(self._surround_command(f"{dep_commands[0]}; {MOCK_COMMAND_LINE}" + RETURN_CODE_CAPTURE))
```
```suggestion
        self.__assert_command_is(self._surround_command(f"{dep_commands[0]}; {MOCK_COMMAND_LINE}{RETURN_CODE_CAPTURE}"))
```

nsoranzo commented on test/unit/app/jobs/test_command_factory.py:
```diff
@@ -37,59 +38,59 @@ def tearDown(self):
         dep_commands = [". /opt/galaxy/tools/bowtie/default/env.sh"]
         self.job_wrapper.dependency_shell_commands = dep_commands
-        self.__assert_command_is(_surround_command(f"{dep_commands[0]}; {MOCK_COMMAND_LINE}; return_code=$?"),
+        self.__assert_command_is(self._surround_command(f"{dep_commands[0]}; {MOCK_COMMAND_LINE}" + RETURN_CODE_CAPTURE),
```
```suggestion
        self.__assert_command_is(self._surround_command(f"{dep_commands[0]}; {MOCK_COMMAND_LINE}{RETURN_CODE_CAPTURE}"),
```

nsoranzo commented on test/unit/app/jobs/test_command_factory.py:
```diff
@@ -37,59 +38,59 @@ def tearDown(self):
         self.include_work_dir_outputs = False
         self.job_wrapper.prepare_input_files_cmds = ["/opt/split1", "/opt/split2"]
-        self.__assert_command_is(_surround_command("/opt/split1; /opt/split2; %s; return_code=$?") % MOCK_COMMAND_LINE)
+        self.__assert_command_is(self._surround_command(f"/opt/split1; /opt/split2; {MOCK_COMMAND_LINE}" + RETURN_CODE_CAPTURE))
```
```suggestion
        self.__assert_command_is(self._surround_command(f"/opt/split1; /opt/split2; {MOCK_COMMAND_LINE}{RETURN_CODE_CAPTURE}"))
```

nsoranzo commented on test/unit/app/jobs/test_command_factory.py:
```diff
@@ -37,59 +38,59 @@ def tearDown(self):
-        self.__assert_command_is(_surround_command(
-            "{} {}/tool_script.sh > ../outputs/tool_stdout 2> ../outputs/tool_stderr; return_code=$?".format(
+        self.__assert_command_is(self._surround_command(
+            "{} {}/tool_script.sh > ../outputs/tool_stdout 2> ../outputs/tool_stderr".format(
```
```suggestion
            "{} {}/tool_script.sh > ../outputs/tool_stdout 2> ../outputs/tool_stderr{}".format(
```

nsoranzo commented on test/unit/app/jobs/test_command_factory.py:
```diff
@@ -37,59 +38,59 @@ def tearDown(self):
                 self.job_wrapper.shell,
                 self.job_wrapper.working_directory,
-            )))
+            ) + RETURN_CODE_CAPTURE))
```
```suggestion
                RETURN_CODE_CAPTURE,
            )))
```

nsoranzo commented on lib/galaxy/jobs/command_factory.py:
```diff
@@ -290,10 +289,12 @@ def capture_stdout_stderr(self, stdout_file, stderr_file):
             self.return_code_captured = True
             self.append_command(CAPTURE_RETURN_CODE)
+            if exit_code_path:
+                self.append_command(f"echo $return_code > {exit_code_path}")
```
Now that this is called only from one place, we can simplify this?

```suggestion
    def capture_return_code(self, exit_code_path):
        self.append_command(CAPTURE_RETURN_CODE)
        self.append_command(f"echo $return_code > {exit_code_path}")
```

nsoranzo commented on lib/galaxy/jobs/command_factory.py:
```diff
@@ -109,12 +109,13 @@ def build_command(
 
     if stdout_file and stderr_file:
         commands_builder.capture_stdout_stderr(stdout_file, stderr_file)
-    commands_builder.capture_return_code()
```
Slightly concerned that the order of these 2 has been reversed, is `capture_stdout_stderr()` now catching the right output? I don't actually understand the Bash code in that method, but maybe a unit test with `stdout_file` and `stderr_file` could help prevent regressions.

mvdbeek replied:
~Either position should be fine.~ This was the idea: https://stackoverflow.com/questions/692000/how-do-i-write-standard-error-to-a-file-while-using-tee-with-a-pipe/692407#692407, and https://github.com/galaxyproject/galaxy/pull/6958#issuecomment-466087700 contains an example of a job script

mvdbeek replied:
But now that I look at it more closely, I don't think it does what I wanted it to do, which is to print to stdout/stderr and to stdout and stderr files, so that job pods could stream the stdout/stderr.
For that I think the right pattern would have been
```diff
diff --git a/lib/galaxy/jobs/command_factory.py b/lib/galaxy/jobs/command_factory.py
index a30ea58338..989ae5f4e4 100644
--- a/lib/galaxy/jobs/command_factory.py
+++ b/lib/galaxy/jobs/command_factory.py
@@ -282,14 +282,14 @@ class CommandsBuilder:
         self.append_command("; ".join(c for c in commands if c))
 
     def capture_stdout_stderr(self, stdout_file, stderr_file):
-        self.prepend_command("""out="${TMPDIR:-/tmp}/out.$$" err="${TMPDIR:-/tmp}/err.$$"
+        self.prepend_command(
+            f"""out="${{TMPDIR:-/tmp}}/out.$$" err="${{TMPDIR:-/tmp}}/err.$$"
 mkfifo "$out" "$err"
 trap 'rm "$out" "$err"' EXIT
-tee -a stdout.log < "$out" &
-tee -a stderr.log < "$err" >&2 &""",
-                             sep="")
-        self.append_command(f"> '{stdout_file}' 2> '{stderr_file}'",
-                            sep="")
+tee -a '{stdout_file}' < "$out" &
+tee -a '{stderr_file}' < "$err" >&2 &""",
+        sep="")
+        self.append_command(f'> "$out" 2> "$err"', sep="")
 
     def capture_return_code(self, exit_code_path):
         self.append_command(CAPTURE_RETURN_CODE)

```

nsoranzo replied:
Thanks, that makes much more sense!

mvdbeek committed:
Simplify capture_return_code

Since we use it from one place only now

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
Use string formatting instead of concatenation

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
One more string formatting

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

qiagu commented on lib/galaxy/jobs/command_factory.py:
```diff
@@ -109,12 +109,13 @@ def build_command(
         commands_builder.prepend_command(container_monitor_command)
 
+    working_directory = remote_job_directory or job_wrapper.working_directory
+    commands_builder.capture_return_code(join(working_directory, f"galaxy_{job_wrapper.job_id}.ec"))
```
Shall we combine utils for the exit_code_file?
https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/job_execution/output_collect.py#L715

mvdbeek committed:
Re-use default_exit_code_file

mvdbeek committed:
Only yield captured retun code when we're capturing the return code

mvdbeek committed:
Move stdout/stderr capture

mvdbeek committed:
One more f-string

mvdbeek committed:
Unit test for named pipe capture

mvdbeek committed:
Fix tee pattern for stdout/stderr streaming

mvdbeek committed:
Extend kubernetes logs to check that job logs are properly recorded

nsoranzo commented on lib/galaxy/jobs/command_factory.py:
```diff
@@ -109,13 +108,12 @@ def build_command(
-    commands_builder.capture_return_code()
-
     if include_metadata and job_wrapper.requires_setting_metadata:
         working_directory = remote_job_directory or job_wrapper.working_directory
```
```suggestion
```

nsoranzo commented on lib/galaxy/jobs/runners/__init__.py:
```diff
@@ -219,12 +219,10 @@ def parse_destination_params(self, params):
+                   job_wrapper,
+                   include_metadata=False,
+                   include_work_dir_outputs=True,
+                   modify_command_for_container=True):
```
Indentation seems off, maybe use the black-complaint style to reduce merge conflicts.

mvdbeek committed:
Fix stdout/stderr capture for container commands

By pulling `capture_stdout_stderr` to the outer scope the capture
runs on the "host" instead of the container.

mvdbeek committed:
Replace container kill trap if present

nsoranzo commented on lib/galaxy/jobs/command_factory.py:
```diff
@@ -92,6 +92,10 @@ def build_command(
         else:
             commands_builder = CommandsBuilder(externalized_commands)
 
+    for_pulsar = 'script_directory' in remote_command_params
```
You can remove an identical line below (line 109).

mvdbeek committed:
Drop redundant statement
